app/src/main/java/openfoodfacts/github/scrachx/openfood/views/splash/SplashActivity.java

```diff
@@ -85,4 +133,237 @@ public void hideLoading(boolean isError) {
+                                        .positiveText("Retry")
+                                        .negativeText(R.string.txtPictureNeededDialogNo)
+                                        .onPositive((dialog, which) -> doDownload())
+                                        .onNegative(new MaterialDialog.SingleButtonCallback() {
+                                            @Override
+                                            public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
+                                                textLoading.setText(getString(R.string.txtLoading));
+                                                presenter.refreshData();
+                                            }
+                                        })
+                                        .show();
+                                Log.d(TAG, "server contact failed");
+                            }
+                        }
+
+                        @Override
+                        public void onFailure(Call<ResponseBody> call, Throwable t) {
+                            new MaterialDialog.Builder(SplashActivity.this)
+                                    .content("Connection failed!")
+                                    .positiveText("Retry")
+                                    .negativeText(R.string.txtPictureNeededDialogNo)
+                                    .onPositive((dialog, which) -> doDownload())
+                                    .onNegative(new MaterialDialog.SingleButtonCallback() {
+                                        @Override
+                                        public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
+                                            textLoading.setText(getString(R.string.txtLoading));
+                                            presenter.refreshData();
+                                        }
+                                    })
+                                    .show();
+                            Log.e(TAG, "error");
+                        }
+                    });
+        } else {
+            new MaterialDialog.Builder(SplashActivity.this)
+                    .content("Wifi Connection not found!")
+                    .positiveText("Retry")
+                    .negativeText(R.string.txtPictureNeededDialogNo)
+                    .onPositive((dialog, which) -> doDownload())
+                    .onNegative(new MaterialDialog.SingleButtonCallback() {
+                        @Override
+                        public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
+                            textLoading.setText(getString(R.string.txtLoading));
+                            presenter.refreshData();
+                        }
+                    })
+                    .show();
+        }
+    }
+
+    public void unzip(ResponseBody responseBody, String _location) {
```

Comment:
Can you make the **unzip** method more like a utility method and move it to Utils.java. This might be used by other developer for other cases.
You can also change its signature to accept an InputStream as a input instead of a ResponseBody and let it throw the exception that can be catch and handled locally when need, in order to make it more generic ex